File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> global = new ArrayList<>();
        Arrays.sort(candidates);
        combinationSum2H(global, new ArrayList<Integer>(), candidates, target, 0, 0);
        return global;
    }

    public void combinationSum2H(
        List<List<Integer>> global,
        List<Integer> res,
        int[] nums,  
        int target, 
        int sum, 
        int index
    ) {
        
        if(sum == target) {
            global.add(new ArrayList<>(res));
            return;
        }
       
        for(int j = index; j < nums.length; j++) {
            if (j > index && nums[j] == nums[j - 1]) continue;

            if(sum + nums[j] > target) {
                continue;
            }
            res.add(nums[j]);
            combinationSum2H(global, res, nums, target, sum + nums[j], j+1);
            res.remove(res.size() - 1);
        }
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for index in range(len(nums)):
            
            if index > 0 and nums[index] == nums[index -1]:
                continue

            target = -nums[index]
            l = index + 1
            r = len(nums) -1

            while l < r:
                if nums[l] + nums[r] == target:
                    res.append([nums[index], nums[l], nums[r]])
                    l += 1
                    r -= 1
                    # ✅ skip duplicates for l
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1

                    # ✅ skip duplicates for r
                    while l < r and nums[r] == nums[r + 1]:
                        r -= 1
    
                elif nums[l] + nums[r] < target:
                    l += 1
                else:
                    r -= 1

        return res     

        